### ISSUE TITLE
USB keyboard does not work when plugged in during power-on

### ISSUE THREAD

**Issue by weizenspreu** [felis/USB_Host_Shield_2.0#68]
For testing purposes I have a Cherry RS 6000 USB keyboard that is connected to the Circuits@Home USB host shield on top of an Arduino Uno R3.

However, when my Arduino device is powered on and the keyboard is already connected, the keyboard just does not work. Only when I disconnect the keyboard from the USB host shield and reconnect it again it starts working.

I used the USBHIDBootKbd example as a starting point which also just works when the keyboard is not connected during power-on.

P.S.: Other keyboards do work even when they connected during power-on, e.g. my Fujitsu-Siemens KB SC SI USB.

**Comment by joseangeljimenez** [comment]
Hello,

I have been able to reproduce the same issue. When powering-up an Arduino with USB Host Shield and a peripheral USB device plugged from the power-up, the library/shield won't detect the device. However, if you then disconnect (unplug) and reconnect (plug) the USB device, the library automatically detects the peripheral.

We have found several USB devices where this happens, among others:
- Kensington Expert Mouse Wired Trackball K64325,
http://www.kensington.com/us/us/4493/k64325/expert-mousespan-classregsymspan-wired-trackball#.VlhjEMrtKzg
- NSI Desktop Office Trackball R60,
http://www.nsi-be.com/pointing-devices/trackball-units/60-mm-desktop-office-trackball/
- Esterline Traxsys DS509162/2 Joystick.

How should I proceed to try to solve this issue? If it helps, we can provide a USBPcap log of the communication with these devices.

Regards,
Jose Angel Jimenez

**Comment by xxxajk** [comment]
Add the USB power switch and the problem should go away.
See this article: https://www.circuitsathome.com/mcu/vbus-power-control-on-usb-host-shield
I use the MIC2004 switch, which resolves all of this sort of problem for me.

**Comment by joseangeljimenez** [comment]
Hello, thanks a lot for the hint. We were already working on this solution, but we thought it was a brute force solution not targetting the root cause of the issue. That is why I also reported our findings here.

So, what is your feeling about this? Should this issue stay open? It should in case you thinkg there is something to fix/improve in the library in order to make this "stubborn" USB devices work without the power cycle. Or do you think, in your experience, that these devices will always need a hard power cycle? In the latter case, why is this issue still open?

Some additional information from my side, in case it is useful for others. You don't neccessarily need an specific power switch IC for power cycling the offending USB device. A simple PNP BJT or p-MOSFET on the high side of the circuit will also work, assuming you are willing to dedicate one pin of the MCU instead of the GPX pin to control the power on-off. With the PNP BJT, you can simply configure your control pin as input (high Z state) in order to switch off the USB device. Or configure the control pin as output LOW to switch it on. I may write a small post on the subject including an schematic if requested.

**Comment by xxxajk** [comment]
The difference is that UHS 2.0.1 (the shield) supports the chip (and a few others) directly. This is to make things easier for you.

**Comment by joseangeljimenez** [comment]
But... should this issue be still open? Is there something to fix/improve in the library?

I am trying to understand, from the USB standards point of view, if these "troublesome" devices are behaving correctly when requiring a power cycle.

Thanks for your support

**Comment by xxxajk** [comment]
Unfortunately, some devices in the wild are broken. How they pass QC is
beyond any idea I have.
We see this mostly with storage devices (thumb drives) however most of
those have been resolved.
The best fix is to power cycle, as it gets you to a state that can be
determined.
Not all devices will reset properly when a burst is sent. While rare, it
happens.


On Tue, Dec 8, 2015 at 6:52 AM, José Ángel Jiménez <notifications@github.com
> wrote:

> But... should this issue be still open? Is there something to fix/improve
> in the library?
>
> I am trying to understand, from the USB standards point of view, if these
> "troublesome" devices are behaving correctly when requiring a power cycle.
>
> Thanks for your support
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/felis/USB_Host_Shield_2.0/issues/68#issuecomment-162860165>
> .
>



-- 
Visit my github for awesome Arduino code @ https://github.com/xxxajk

**Comment by joseangeljimenez** [comment]
Your experience on this is very helpful. Thank you. I will test again the problematic devices, now with the hardware power cycle in place, and report back the results.

**Comment by Grunewald17** [comment]
Im using a Arduino Mega ADK Board with Ethernet Shield. This issue exist only if I use Keyboard and Mouse at the same time with an USB Hub. But, it's working if you plug the keyboard in the right USB Hub port. USB Hub Port 2 isn't working, so I have to plug the keyboard into USB Hub Port 1 and the USB Mouse into Port 2. Then you don't need the VBus power trick (this trick was not working for me).

It would be nice if someone would write a bugfree USB Shield driver to use USB Mouse, USB Keyboard, USB Joystick and USB Midi at the same time with an USB Hub for the Arduino Mega ADK Board.

**Comment by xxxajk** [comment]
Do you include the hub code?

On Thu, Dec 10, 2015 at 4:18 AM, Grunewald17 <notifications@github.com>
wrote:

> Im using a Arduino Mega ADK Board with Ethernet Shield. This issue exist
> only if I use Keyboard and Mouse at the same time with an USB Hub. But,
> it's working if you plug the keyboard in the right USB Hub port. USB Hub
> Port 2 isn't working, so I have to plug the keyboard into USB Hub Port 1
> and the USB Mouse into Port 2. Then you don't need the VBus power trick
> (this trick was not working for me).
>
> It would be nice if someone would write a bugfree USB Shield driver to use
> USB Mouse, USB Keyboard, USB Joystick and USB Midi at the same time with an
> USB Hub for the Arduino Mega ADK Board.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/felis/USB_Host_Shield_2.0/issues/68#issuecomment-163541284>
> .
>



-- 
Visit my github for awesome Arduino code @ https://github.com/xxxajk

**Comment by Grunewald17** [comment]
Yes. Also the Mouse Wheel is not working with 99% of tested mouse devices. It would be nice if there were a solution for this.

**Comment by xxxajk** [comment]
Mouse wheel is not supported with HID BOOT because that is not in the HID
BOOT specification.
Many mice should be able to be used via a "HID RAW" style driver. If it is,
then you need to figure out what the messages returned actually mean. The
problem with mice is that there tends to be many different ones supporting
extra buttons, wheels, etc, and there is no "standard" to go by in order to
talk to them. HID BOOT should work with most of them.

I am unaware of any mouse that does not support HID BOOT at a minimum. Try
with no hub, and only the device you are using, and only the code to
support that device only.... If it works, then read on...

Part of the issue could be in how UHS20 enumerates. If this is partly the
issue (which I suspect it is!) UHS30 (when it supports everything and more)
won't have this issue since it enumerates as per the actual specification,
instead of the UHS20 hit-and-miss method, which can also lock some devices
up. Hard to explain without going into deep details, but it is fair to say
that UHS20 does not do things totally right during enumeration and can be
position dependent when multiple devices are used. It is a known limitation
and issue with the implementation, which I tried to improve a little bit,
but abandoned because, well, it is broken.






On Fri, Dec 11, 2015 at 8:33 AM, Grunewald17 <notifications@github.com>
wrote:

> Yes. Also the Mouse Wheel is not working with 99% of tested mouse devices.
> It would be nice if there were a solution for this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/felis/USB_Host_Shield_2.0/issues/68#issuecomment-163937087>
> .
>



-- 
Visit my github for awesome Arduino code @ https://github.com/xxxajk

**Comment by Grunewald17** [comment]
I've tested the last UHS30 ... the same issues.

**Comment by xxxajk** [comment]
UHS30 does not have HID support.

On Fri, Dec 11, 2015 at 12:03 PM, Grunewald17 <notifications@github.com>
wrote:

> I've tested the last UHS30 ... the same issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/felis/USB_Host_Shield_2.0/issues/68#issuecomment-163991125>
> .
>



-- 
Visit my github for awesome Arduino code @ https://github.com/xxxajk

**Comment by Grunewald17** [comment]
The beta UHS30 ?

**Comment by xxxajk** [comment]
Correct. All that is supported is hub and mass storage at present.
